### DISCUSSION TITLE
Nested filter AND concatination

### DISCUSSION THREAD

**Discussion by mxcd** [pocketbase/pocketbase#5290] [Q&A] [answered]
<img width="1416" alt="image" src="https://github.com/user-attachments/assets/2103ba1b-f2c4-4799-806e-1c57f887733e">
  
I have the above data structure implemented in my pocketbase project.  
My goal is to be able to filter `images` on one or more `image_tags` with an `AND` concatination  
```
"List all images that have image_tag X AND image_tag Y assigned to it"
```
  
So far I have only been able to get a filter working that allows filtering for a single `image_tag`:
```
const result = await pb.collection<ImageWithTagsType>("images").getList(0, 1000, {
  filter: "(image_tag_assignments_via_image.imageTag?='qtzb27wceinc1x3')",
  sort: "-capturedAtCorrected",
  expand: "camera, project, image_tag_assignments_via_image, image_tag_assignments_via_image.imageTag",
});
```  
  
Using a filter like the one below leads to an empty result set:
```
(image_tag_assignments_via_image.imageTag?='qtzb27wceinc1x3' && image_tag_assignments_via_image.imageTag?='89ewmdtae932h0r')
```  
  
Any suggestions?

**Comment by ganigeorgiev** [comment, marked as answer]
Your problem seems similar to https://github.com/pocketbase/pocketbase/discussions/4319, aka. both `image_tag_assignments_via_image.imageTag` conditions refer to a single `imageTag` row-column value and it cannot be 2 different things at the same time.

The suggested solution in the above linked discussion will not work in your case because you have a junction collection (even with a custom route and raw SQL I'm not sure if it can be done directly unless you join multiple times the `image_tag_assignments` for the number of tags you want to check with `AND`).

If you don't want to change your structure to have a multipart `relation` field, one possible workaround could be to create a helper view collection that will aggregate and serialize all image tags in one place as json array and then you can search for the specific id with the `~` operator. Something like:

```sql
# serialized_image_tags
SELECT
    images.id as id,
    images.id as image, # duplicate so that it can be used as relation
    json_group_array(ita.imageTag) as imageTags
FROM images
LEFT JOIN image_tag_assignments ita on ita.image = images.id
GROUP BY images.id
```

Then your request could look like:

```ts
const result = await pb.collection<ImageWithTagsType>("images").getList(1, 1000, {
  filter: `serialized_image_tags_via_image.imageTags ?~ '"qtzb27wceinc1x3"' && serialized_image_tags_via_image.imageTags ?~ '"89ewmdtae932h0r"')`,
  sort: "-capturedAtCorrected",
  expand: "camera, project, image_tag_assignments_via_image.imageTag",
});
```

_edit: added missing `GROUP BY` to the view collections query_

**Reply by mxcd** [reply]
Hi @ganigeorgiev  
Thank you so much for your suggestion!  
I think it is almost working but I think the view collection is not correct:  
<img width="1353" alt="image" src="https://github.com/user-attachments/assets/38e19633-4b84-4bdb-819c-81e26ea472bb">
<img width="276" alt="image" src="https://github.com/user-attachments/assets/c7a20503-9f21-4b56-b1ce-a4970afef958">
The view collection only holds a single image which has a json array holding all 16k `image_tags` IDs that are mapped my `image_tag_assignments`

**Reply by ganigeorgiev** [reply]
Yes, sorry. The `GROUP BY` statement is missing, I've updated the answer.

**Reply by mxcd** [reply]
Thank you! This worked now.
BUT: The request now takes approx. 5 seconds (with 3k images and 16k image_tag_assignments in the DB) 🫠    
Any way to create an index on the serialized_image_tags or something to speed this up?

**Reply by mxcd** [reply]
Maybe, as an alternative solution, I could simply keep the image_tags relation redundantly also in the image object and update it with a hook upon image_tag_assignment creation or deletion. Filtering directly on images for multiple image_tags should work, right?  
__
I need the `image_tag_assignments` to store additional metadata for the assignment, but my filtering does not care about those additional information

**Reply by ganigeorgiev** [reply]
> Any way to create an index on the serialized_image_tags or something to speed this up?

No, views cannot have indexes. You can try to add individual indexes on the `image_tag_assignments.image` and `image_tag_assignments.imageTag` fields and see if the query planner will take a different route. The sort field (`capturedAtCorrected`) may also benefit from an index.

> Maybe, as an alternative solution, I could simply keep the image_tags relation redundantly also in the image object...

Yes, that should work.

**Reply by mxcd** [reply]
The index didn't help unfortunately. I'll try the additional relation workaround. Anyway - your answer is correct and works (maybe just not for bigger datasets). Thank you very much!
